File: src/GitRemoteS3/Git.hs
-- SPDX-FileCopyrightText: 2023-present Amazon.com, Inc. or its affiliates
--
-- SPDX-License-Identifier: Apache-2.0

-- | Wrappers around @git@ subprocess calls.
--
-- Mirrors @git_remote_s3/git.py@.
module GitRemoteS3.Git
  ( GitError (..)
  , archive
  , bundle
  , unbundle
  , revParse
  , isAncestor
  , getRemoteUrl
  , validateRefName
  , getLastCommitMessage
  ) where

import           Control.Exception          (Exception, throwIO)
import           System.Exit                (ExitCode (..))
import           System.IO                  (stderr, hPutStr)
import           System.Process             ( readProcessWithExitCode
                                            , createProcess
                                            , proc
                                            , std_out
                                            , std_err
                                            , StdStream (..)
                                            , waitForProcess
                                            )
import           Text.Regex.TDFA            ((=~))

-- ---------------------------------------------------------------------------
-- Exception type
-- ---------------------------------------------------------------------------

-- | Raised when a @git@ subprocess call fails.
newtype GitError = GitError String
  deriving (Show)

instance Exception GitError

-- ---------------------------------------------------------------------------
-- Helpers
-- ---------------------------------------------------------------------------

-- | Run a @git@ command, capturing stdout and stderr.
-- Throws 'GitError' on non-zero exit code.
runGit :: [String] -> IO String
runGit args = do
  (code, out, err) <- readProcessWithExitCode "git" args ""
  case code of
    ExitSuccess   -> return out
    ExitFailure _ -> throwIO (GitError err)

-- ---------------------------------------------------------------------------
-- Public API
-- ---------------------------------------------------------------------------

-- | @git archive --format zip --output \<folder\>\/repo.zip \<ref\>@
--
-- Returns the path to the created archive file.
archive :: FilePath  -- ^ Destination folder
        -> String    -- ^ Git ref to archive
        -> IO FilePath
archive folder ref = do
  let filePath = folder ++ "/repo.zip"
  _ <- runGit ["archive", "--format", "zip", "--output", filePath, ref]
  return filePath

-- | @git bundle create \<folder\>\/\<sha\>.bundle \<ref\>@
--
-- Returns the path to the created bundle file.
bundle :: FilePath  -- ^ Destination folder
       -> String    -- ^ SHA (used as the bundle file name stem)
       -> String    -- ^ Git ref to bundle
       -> IO FilePath
bundle folder sha ref = do
  let filePath = folder ++ "/" ++ sha ++ ".bundle"
  _ <- runGit ["bundle", "create", filePath, ref]
  return filePath

-- | @git bundle unbundle \<folder\>\/\<sha\>.bundle \<ref\>@
--
-- stdout is redirected to stderr so git's progress output reaches the
-- user (mirrors the Python implementation which sets @stdout=sys.stderr@).
unbundle :: FilePath  -- ^ Folder containing the bundle
         -> String    -- ^ SHA (bundle file name stem)
         -> String    -- ^ Git ref
         -> IO ()
unbundle folder sha ref = do
  let bundlePath = folder ++ "/" ++ sha ++ ".bundle"
  (_, _, _, ph) <- createProcess
    (proc "git" ["bundle", "unbundle", bundlePath, ref])
      { std_out = UseHandle stderr
      , std_err = UseHandle stderr
      }
  code <- waitForProcess ph
  case code of
    ExitSuccess   -> return ()
    ExitFailure _ -> throwIO (GitError "git bundle unbundle failed")

-- | @git rev-parse \<ref\>@
--
-- Returns the full SHA for the given ref.
-- Throws 'GitError' when the ref is not found.
revParse :: String -> IO String
revParse ref = do
  (code, out, _) <- readProcessWithExitCode "git" ["rev-parse", ref] ""
  case code of
    ExitSuccess   -> return (strip out)
    ExitFailure _ -> throwIO (GitError ("fatal: " ++ ref ++ " not found"))
  where
    strip = reverse . dropWhile (== '\n') . reverse

-- | @git merge-base --is-ancestor \<ancestor\> \<descendant\>@
--
-- Returns @True@ iff @ancestor@ is an ancestor of @descendant@.
isAncestor :: String -> String -> IO Bool
isAncestor ancestor descendant = do
  (code, _, _) <- readProcessWithExitCode
    "git" ["merge-base", "--is-ancestor", ancestor, descendant] ""
  return (code == ExitSuccess)

-- | @git remote get-url \<remote\>@
--
-- Returns the URL configured for the named remote.
-- Throws 'GitError' when the remote is not found.
getRemoteUrl :: String -> IO String
getRemoteUrl remote = do
  (code, out, _) <- readProcessWithExitCode "git" ["remote", "get-url", remote] ""
  case code of
    ExitSuccess   -> return (strip out)
    ExitFailure _ -> throwIO (GitError ("fatal: " ++ remote ++ " not found"))
  where
    strip = reverse . dropWhile (== '\n') . reverse

-- | Validate a git ref name according to git's own rules.
--
-- Returns @True@ iff the name is valid.
-- Mirrors the regex from @refs.c@ in the git source tree:
-- https://github.com/git/git/blob/406f326d/refs.c#L170
validateRefName :: String -> Bool
validateRefName name =
  not (name =~ badPattern :: Bool)
  where
    badPattern :: String
    badPattern =
      "(^\\.)|(\\.\\.)|([:?\\[\\\\\\^~\\s\\*]])|\\.lock$|(/$)|(@\\{)|([\\x00-\\x1f])"

-- | @git log -1 --pretty=%h %s@
--
-- Returns a short summary of the last commit.
-- Throws 'GitError' on failure.
getLastCommitMessage :: IO String
getLastCommitMessage = do
  (code, out, _) <- readProcessWithExitCode "git" ["log", "-1", "--pretty=%h %s"] ""
  case code of
    ExitSuccess   -> return (strip out)
    ExitFailure _ -> throwIO (GitError "fatal: an error has occurred")
  where
    strip = reverse . dropWhile (== '\n') . reverse

File: src/GitRemoteS3/Git.php
<?php

// SPDX-FileCopyrightText: 2023-present Amazon.com, Inc. or its affiliates
//
// SPDX-License-Identifier: Apache-2.0

declare(strict_types=1);

namespace GitRemoteS3;

/**
 * Exception thrown when a git sub-process fails.
 *
 * Corresponds to GitError in git_remote_s3/git.py.
 */
class GitError extends \RuntimeException {}

/**
 * Helpers that wrap git sub-processes.
 *
 * All methods are static; they spawn git via proc_open / exec and throw
 * GitError on failure — mirroring the Python module git_remote_s3/git.py.
 */
class Git
{
    /**
     * Archive the repository into a repo.zip file.
     *
     * @param string $folder  Directory in which to write the archive
     * @param string $ref     Git ref to archive
     * @return string         Absolute path to the written archive
     * @throws GitError
     */
    public static function archive(string $folder, string $ref): string
    {
        $filePath = "{$folder}/repo.zip";
        [$stdout, $stderr, $code] = self::run(
            ['git', 'archive', '--format', 'zip', '--output', $filePath, $ref]
        );
        if ($code !== 0) {
            throw new GitError($stderr);
        }
        return $filePath;
    }

    /**
     * Bundle a git ref into <sha>.bundle.
     *
     * @param string $folder  Directory in which to write the bundle
     * @param string $sha     SHA to use as the filename (without extension)
     * @param string $ref     Git ref to bundle
     * @return string         Absolute path to the bundle file
     * @throws GitError
     */
    public static function bundle(string $folder, string $sha, string $ref): string
    {
        $filePath = "{$folder}/{$sha}.bundle";
        [$stdout, $stderr, $code] = self::run(
            ['git', 'bundle', 'create', $filePath, $ref]
        );
        if ($code !== 0) {
            throw new GitError($stderr);
        }
        return $filePath;
    }

    /**
     * Unbundle a <sha>.bundle file into the current repository.
     *
     * @param string $folder  Directory that contains the bundle
     * @param string $sha     SHA portion of the bundle filename
     * @param string $ref     Git ref to check out after unbundling
     * @throws GitError
     */
    public static function unbundle(string $folder, string $sha, string $ref): void
    {
        // Stderr is forwarded to the process's own stderr (like Python's stdout=sys.stderr).
        $cmd = sprintf(
            'git bundle unbundle %s %s 1>&2',
            escapeshellarg("{$folder}/{$sha}.bundle"),
            escapeshellarg($ref)
        );
        passthru($cmd, $code);
        if ($code !== 0) {
            throw new GitError("git bundle unbundle exited with code {$code}");
        }
    }

    /**
     * Resolve a git ref to its full SHA.
     *
     * @param string $ref
     * @return string  40-character hex SHA
     * @throws GitError
     */
    public static function revParse(string $ref): string
    {
        [$stdout, $stderr, $code] = self::run(['git', 'rev-parse', $ref]);
        if ($code !== 0) {
            throw new GitError("fatal: {$ref} not found");
        }
        return trim($stdout);
    }

    /**
     * Check whether $ancestor is an ancestor of $descendant.
     *
     * @param string $ancestor
     * @param string $descendant
     * @return bool
     */
    public static function isAncestor(string $ancestor, string $descendant): bool
    {
        [$stdout, $stderr, $code] = self::run(
            ['git', 'merge-base', '--is-ancestor', $ancestor, $descendant],
            suppressOutput: true
        );
        return $code === 0;
    }

    /**
     * Return the URL configured for a named remote.
     *
     * @param string $remote
     * @return string
     * @throws GitError
     */
    public static function getRemoteUrl(string $remote): string
    {
        [$stdout, $stderr, $code] = self::run(['git', 'remote', 'get-url', $remote]);
        if ($code !== 0) {
            throw new GitError("fatal: {$remote} not found");
        }
        return trim($stdout);
    }

    /**
     * Validate a git refname according to the rules in git's refs.c.
     *
     * @see https://github.com/git/git/blob/406f326/refs.c#L170
     *
     * @param string $name
     * @return bool
     */
    public static function validateRefName(string $name): bool
    {
        return preg_match(
            '/(\.\.)|(^\.)|([:?\[\\\\^~\s*])|(\.lock$)|(\/\/)|(^\/)|(\/$)|(@\{)|([\x00-\x1f])/',
            $name
        ) === 0;
    }

    /**
     * Return a short summary of the last commit (hash + subject).
     *
     * @return string
     * @throws GitError
     */
    public static function getLastCommitMessage(): string
    {
        [$stdout, $stderr, $code] = self::run(['git', 'log', '-1', '--pretty=%h %s']);
        if ($code !== 0) {
            throw new GitError('fatal: an error has occurred');
        }
        return trim($stdout);
    }

    // -------------------------------------------------------------------------
    // Internal helper
    // -------------------------------------------------------------------------

    /**
     * Run an external command and capture stdout + stderr.
     *
     * @param string[]   $argv
     * @param bool       $suppressOutput  When true, stderr/stdout are discarded
     * @return array{string, string, int}  [stdout, stderr, exit_code]
     */
    private static function run(array $argv, bool $suppressOutput = false): array
    {
        $cmd = implode(' ', array_map('escapeshellarg', $argv));

        $descriptors = [
            0 => ['pipe', 'r'],  // stdin
            1 => ['pipe', 'w'],  // stdout
            2 => ['pipe', 'w'],  // stderr
        ];

        $process = proc_open($cmd, $descriptors, $pipes);
        if (!is_resource($process)) {
            throw new GitError("Failed to start process: {$cmd}");
        }

        fclose($pipes[0]);

        $stdout = stream_get_contents($pipes[1]);
        $stderr = stream_get_contents($pipes[2]);
        fclose($pipes[1]);
        fclose($pipes[2]);

        $code = proc_close($process);

        return [$stdout ?: '', $stderr ?: '', $code];
    }
}
